src/services/user.service.js

```
@@ -0,0 +1,200 @@
+const { ApiError } = require('../exeptions/api.error');
+const bcrypt = require('bcrypt');
+const { Token } = require('../models/token.js');
+
+function getAllActivated() {
+  return User.findAll({
+    where: {
+      activationToken: null,
+    },
+  });
+}
+
+function normalize({ id, email, name }) {
+  return { id, email, name };
+}
+
+function findByEmail(email) {
+  return User.findOne({ where: { email } });
+}
+
+async function updateName(name, userId) {
+  const user = await User.findByPk(userId);
+
+  if (!user) {
+    throw ApiError.badRequest('User not found');
+  }
+
+  user.name = name;
+
+  await user.save();
+
+  return user;
+}
+
+async function updatePassword(oldPassword, newPassword, userId) {
+  const user = await User.findByPk(userId);
+
+  if (!user) {
+    throw ApiError.notFound('User not found');
+  }
+
+  const isOldPasswordValid = await bcrypt.compare(oldPassword, user.password);
+
+  if (!isOldPasswordValid) {
+    throw ApiError.badRequest('Old password is incorrect');
+  }
+
+  user.password = await bcrypt.hash(newPassword, 10);
+  await user.save();
+
+  return user;
+}
+
+async function updateEmailRequest(password, userId, newEmail) {
+  const user = await User.findByPk(userId);
+
+  if (!user) {
+    throw ApiError.notFound('User not found');
+  }
+
+  const isPasswordValid = await bcrypt.compare(password, user.password);
+
+  if (!isPasswordValid) {
+    throw ApiError.badRequest('Password is incorrect');
+  }
+
+  const token = await Token.findOne({ where: { userId } });
+
+  const emailChangeToken = uuidv4();
+
+  if (token) {
+    token.emailChangeToken = emailChangeToken;
+    await token.save();
+  } else {
+    await Token.create({
+      userId,
+      emailChangeToken,
+    });
+  }
+
+  user.newEmail = newEmail;
+  await user.save();
+
+  await emailService.sendEmailChangeToken(user.email, emailChangeToken);
+
+  return { message: 'На стару пошту надіслано листа з підтвердженням' };
```

Minor suggestion: `updateEmailRequest` relies on the controller to validate `newEmail`. Consider validating the new email again in the service (or explicitly documenting that controller validation is required) and possibly checking uniqueness before saving `user.newEmail` so you can return a helpful error instead of a DB constraint failure later. This improves robustness of the email-change flow and user experience .